### PR TITLE
fix(cloud_functions,web): use `kIsWeb` to check platform is Web

### PR DESCRIPTION
**Pull request by MichaelDark**
## `defaultTargetPlatform` does not allow to use Cloud Functions emulator in Web integration tests 

In tests `defaultTargetPlatform` always returns `TargetPlatform.android` which does not allow to use `useFunctionsEmulator` in Web integration tests (`localhost` string in url was always replaced by `10.0.2.2`). 
Now it is replaced with native implementation (without usage of `defaultTargetPlatform`).

## Related Issues

No issues were found and/or filed yet, proposing direct solution.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by russellwheatley** [comment]
I'm not sure on what the problem is exactly.

I've just printed out the `defaultTargetPlatform` for each environment:

web (running chromedriver on macOS):
```bash
[LOG]: Platform: TargetPlatform.macOS
```

android
```bash
[LOG]: Platform: TargetPlatform.android
```

iOS
```bash
[LOG]: Platform: TargetPlatform.iOS
```

which is consistent with the [documentation](https://api.flutter.dev/flutter/foundation/defaultTargetPlatform.html).

Is there something I'm missing, @MichaelDark?

**Comment by MichaelDark** [comment]
@russellwheatley 
According to the [documentation](https://api.flutter.dev/flutter/foundation/defaultTargetPlatform.html), for test environments it always returns `Platform.android`, thats why our CI failed with Flutter Web driver tests. Because URL had been replacing to `10.0.2.2` even for Web driver tests. And as far as I know, there is currently no way to override target platform to Web.

> In a test environment, the platform returned is TargetPlatform.android regardless of the host platform.

And thats why I stuck with `cloud_functions` version upgrade (as a result - our project could not be upgraded to latest Firebase ecosystem version. The only workaround for now - to override dependency to this PRs reference)

```
dependencies:
  # ...
  cloud_functions:
    git: 
      url: https://github.com/MichaelDark/flutterfire.git
      ref: fix/cloud-functions-fix-test-target
      path: packages/cloud_functions/cloud_functions
  cloud_firestore: ^1.0.1
  firebase_core: ^1.0.1
```

**Comment by Salakar** [comment]
Can we not switch to using `kIsWeb` to simplify this? 

```dart
import 'package:flutter/foundation.dart' show kIsWeb;

// ...
if (!kIsWeb && isPlatformAndroid) {
  // ...
}
```

**Comment by MichaelDark** [comment]
@Salakar @russellwheatley 
1) We cannot just use `kIsWeb` with `Platform.isAndroid` because:
- `dart:io` is not supported on Web and this code will not compile;
- conditional import is required in order to keep code compilable both for web and other platforms (which support `dart:io`);
- `dart:html` also has `Platform` class, but it is completely not the same as `Platform` from `dart:io`, it cannot be reused to check `Platform.isAndroid`.

2) Current driver tests succeed because we are testing `useFunctionsEmulator` with `0.0.0.0` emulator URL, while under the hood `cloud_function#useFunctionsEmulator` handle only `127.0.0.1` and `localhost` hostnames. If we change right now `0.0.0.0` to `localhost`/`127.0.0.1` in driver tests - CI will fail.

**Comment by rrousselGit** [comment]
> dart:io is not supported on Web and this code will not compile;

Can you expand on this? `kIsWeb` being a constant, tree-shaking should be able to detect that io is unused

**Comment by Salakar** [comment]
> @Salakar @russellwheatley
> 
> 1. We cannot just use `kIsWeb` with `Platform.isAndroid` because:
> 
> * `dart:io` is not supported on Web and this code will not compile;

To add to previous comment, my understanding is that [kIsWeb](https://api.flutter.dev/flutter/foundation/kIsWeb-constant.html) does not depend on `dart:io`, in fact we use it all over our web integration & driver tests without compilation issues 🤔 as for the `Platform.isAndroid` one you mention we don't use that either, we use `TargetPlatform` & `defaultTargetPlatform` from Flutter which doesn't use io on web?

**Comment by rrousselGit** [comment]
In any case, it would be valuable to add a test for the changed behavior.

This would both prevent regressions and help us understand what this PR is trying to do.

**Comment by MichaelDark** [comment]
@rrousselGit Sure, I will try to develop appropriate driver tests. I will ask you when I will need any help 😉

**Comment by MichaelDark** [comment]
@rrousselGit @russellwheatley @Salakar  Maybe it is even better to pass as a parameter `{bool isAndroid}` and delegate this case handling to end users?

**Comment by MichaelDark** [comment]
@rrousselGit In addition to message above, it would be easier to test this optional Android-case-handling and it will be much more straightforward to use (FYI this bug delayed our project dependencies upgrade for more than a month 😢 )

File: packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart
Comment (rrousselGit) [inline packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:76]: What about an assert?

Like:

```dart
assert(
  kIsWeb && defaultTargetPlatform != TargetPlatform.web, 
  '''
Tried to use `useFunctionsEmultator` on web, but `defaultTargetPlatform` indicated that the platform is android`.

This can happen inside tests when <add explanation>.
To fix, consider using `debugDefaultTargetPlatformOverride` like so:

void main() {
  debugDefaultTargetPlatformOverride = TargetPlatform.web;
  test('some test', () {...})
}
```

Comment (MichaelDark) [reply packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:76]: @rrousselGit No `TargetPlatform.web` in stable :(

Comment (rrousselGit) [reply packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:76]: Hum that's a shame, I don't mind then!

Comment (MichaelDark) [reply packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:76]: @rrousselGit The only way to indicate whether it is Web or not - `kIsWeb`, and currently there is not `TargetPlatform.web` in Flutter (stable channel)

File: packages/cloud_functions/cloud_functions/test/firebase_functions_test.dart
Comment (rrousselGit) [inline packages/cloud_functions/cloud_functions/test/firebase_functions_test.dart:137]: prefer `for (... in)` instead

Comment (MichaelDark) [reply packages/cloud_functions/cloud_functions/test/firebase_functions_test.dart:137]: @rrousselGit `for (var ...)` or `for (TargetPlatform ...)`?

Comment (rrousselGit) [reply packages/cloud_functions/cloud_functions/test/firebase_functions_test.dart:137]:
```dart
for (final platform in TargetPlatform.values) {
 ...
}
```

And same thing for `testLocalhostOrigins`

**Comment by MichaelDark** [comment]
@rrousselGit Committed changes, also move localhost hostnames to single place (with `@visibleForTesting` annotation). Waiting for review and GH Actions to pass

File: packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart
Comment (rrousselGit) [inline packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:88]: I personally prefer the previous code

To me this feels too abstracted.

Comment (MichaelDark) [reply packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:88]: @rrousselGit As you wish, you are the power here, just reverted this changes 👌🏻

File: packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart
Comment (rrousselGit) [inline packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:95]:
```suggestion
      if (origin.startsWith('http://localhost')) {
        _origin = origin.replaceFirst('http://localhost', 'http://10.0.2.2');
        return;
      }
      if (origin.startsWith('http://127.0.0.1')) {
        _origin = origin.replaceFirst('http://127.0.0.1', 'http://10.0.2.2');
        return;
```

File: packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart
Comment (rrousselGit) [inline packages/cloud_functions/cloud_functions/lib/src/firebase_functions.dart:78]:
```suggestion
```

**Review by rrousselGit — APPROVED** [review]

**Comment by rrousselGit** [comment]
Looking good, thanks!

**Review by Salakar — APPROVED** [review]

**Comment by rrousselGit** [comment]
I'll wait for the CI just in case, then merge it.

**Comment by Salakar** [comment]
@MichaelDark thanks a lot for this PR and addressing all the feedback. 🎉

**Comment by MichaelDark** [comment]
@rrousselGit Where I can read about expectations when it will be released after merge to master?

**Comment by rrousselGit** [comment]
@Salakar is probably better placed to answer this

**Comment by Salakar** [comment]
> @rrousselGit Where I can read about expectations when it will be released after merge to master?

Still to be documented but TLDR; we release on Monday's. But little fixes like this we can just ship whenever so if you need it now I'm sure we can ship it for you.

**Comment by MichaelDark** [comment]
@Salakar Thank you for the extra information! Looking forward to see new version in `pub.dev`

**Comment by Salakar** [comment]
> @Salakar Thank you for the extra information! Looking forward to see new version in `pub.dev`

The CI gods have finally let us through so have merged now. Will ship once I'm back at PC, in an hour or so.

**Comment by MichaelDark** [comment]
@Salakar Good day, just reminding you about the release 👍🏻

**Comment by Salakar** [comment]
@MichaelDark This has just been released by @rrousselGit, apologies for the delay - had some other things we wanted to ship at the same time.

**Pull request merged**
